src/greynirseq/nicenlp/models/simple_parser.py

```diff
@@ -0,0 +1,246 @@
+import itertools
+import logging
+
+
+import torch
+from torch import nn
+import torch.nn.functional as F
+from torch.nn.utils.rnn import pad_sequence
+
+from fairseq import utils
+from fairseq.models.roberta.model import base_architecture, RobertaModel
+from fairseq.models.roberta.hub_interface import RobertaHubInterface
+from fairseq.models import register_model, register_model_architecture
+from fairseq.modules import LayerNorm
+from fairseq.data import (
+    RightPadDataset,
+    BaseWrapperDataset,
+    NestedDictionaryDataset,
+    NumelDataset,
+)
+from fairseq.models.roberta.model import RobertaEncoder
+
+from greynirseq.nicenlp.data.datasets import (
+    WordEndMaskDataset,
+    LabelledSpanDataset,
+    ProductSpanDataset,
+    NumSpanDataset,
+    NestedDictionaryDatasetFix,
+    NestedDictionaryDatasetFix2,
+)
+
+from greynirseq.nicenlp.utils.label_schema.label_schema import make_vec_idx_to_dict_idx
+
+from greynirseq.nicenlp.utils.constituency.greynir_utils import Node
+import greynirseq.nicenlp.utils.constituency.greynir_utils as greynir_utils
+
+import greynirseq.nicenlp.chart_parser as chart_parser
+
+
+logger = logging.getLogger(__name__)
+
+
+class ChartParserHead(nn.Module):
+    def __init__(self, in_features, out_features, dropout):
+        super().__init__()
+        self.in_features = in_features
+        self.inner_dim = in_features
+        self.out_features = out_features
+
+        self.dropout = nn.Dropout(p=dropout)
+        self.layer_norm = LayerNorm(self.inner_dim)
+        self.activation_fn = utils.get_activation_fn("relu")
+
+        self.dense = nn.Linear(self.in_features, self.inner_dim)
+        self.dense2 = nn.Linear(self.inner_dim, self.inner_dim)
+        self.out_proj = nn.Linear(self.inner_dim, self.out_features - 1)
+
+    def forward(self, features, **kwargs):
+        x = self.dense(features)
+        x = self.dropout(x)
+        x = self.layer_norm(x)
+        x = self.activation_fn(x)
+
+        # This constructs all spans via broadcasting
+        # (B x T1 x T2 x C) <- (B x T1 x C) - (B x T2 x C)
+        # bijf <- bif - bjf
+        x = x.unsqueeze(1) - x.unsqueeze(2)
+
+        x = self.dense2(x)
+        x = self.activation_fn(x)
+
+        logits = self.out_proj(x)
+
+        bsz, nwords, _, _ = logits.shape
+        # logits: (bsz, nwords * nwords, nlabels-1)
+        # make label_idx of 0 correspond to NULL label, fixed at label-span score of 0
+        logits = torch.cat([logits.new_zeros(bsz, nwords, nwords, 1), logits], -1)
+
+        return logits
+
+
+@register_model("icebert_simple_parser")
+class SimpleParserModel(RobertaModel):
+    """Simple chart parser model based on Roberta. Simple in the sense that it does not
+       add additional self-attentive layers or feature-engineering from Kitaev and Klein 2018."""
+
+    def __init__(self, args, encoder, task):
+        super().__init__(args, encoder)
+
+        def freeze_module_params(m):
+            if m is not None:
```

Comment:
Do we really want to continue if m is None?

Reply:
There is nothing else in the stub after the if statement.